lara-typescript/src/interactive-api-client/hooks.ts

```diff
@@ -0,0 +1,86 @@
+    return () => {
+      client.removeInteractiveStateListener<InteractiveState>(handleStateUpdate);
+    };
+  }, []);
+
+  const handleSetInteractiveState = (state: InteractiveState) => {
+    setInteractiveState(state);
+    client.setInteractiveState<InteractiveState>(state);
+  };
+
+  return { interactiveState, setInteractiveState: handleSetInteractiveState };
+};
+
+export const useAuthoredState = <AuthoredState>() => {
+  const [ authoredState, setAuthoredState ] = useState<AuthoredState | null>(client.getAuthoredState<AuthoredState>());
+
+  useEffect(() => {
+    // Setup client event listeners. They will ensure that another instance of this hook (or anything else
+    // using client directly) makes changes to authored state, this hook will receive these changes.
+    const handleStateUpdate = (newState: AuthoredState) => {
+      setAuthoredState(newState);
+    };
+    client.addAuthoredStateListener<AuthoredState>(handleStateUpdate);
+    return () => {
+      client.removeAuthoredStateListener<AuthoredState>(handleStateUpdate);
+    };
+  }, []);
+
+  const handleSetAuthoredState = (state: AuthoredState) => {
+    setAuthoredState(state);
+    client.setAuthoredState<AuthoredState>(state);
+  };
+
+  return { authoredState, setAuthoredState: handleSetAuthoredState };
+};
+
+export const useGlobalInteractiveState = <GlobalInteractiveState>() => {
+  const [ globalInteractiveState, setGlobalInteractiveState ] = useState<GlobalInteractiveState | null>(client.getGlobalInteractiveState<GlobalInteractiveState>());
+
+  useEffect(() => {
+    // Setup client event listeners. They will ensure that another instance of this hook (or anything else
+    // using client directly) makes changes to global interactive state, this hook will receive these changes.
+    const handleStateUpdate = (newState: GlobalInteractiveState) => {
+      setGlobalInteractiveState(newState);
+    };
+    client.addGlobalInteractiveStateListener<GlobalInteractiveState>(handleStateUpdate);
+    return () => {
+      client.removeGlobalInteractiveStateListener<GlobalInteractiveState>(handleStateUpdate);
+    };
+  }, []);
+
+  const handleSetGlobalInteractiveState = (state: GlobalInteractiveState) => {
+    setGlobalInteractiveState(state);
+    client.setGlobalInteractiveState<GlobalInteractiveState>(state);
+  };
+
+  return { globalInteractiveState, setGlobalInteractiveState: handleSetGlobalInteractiveState };
+};
+
+export const useInitMessage = <InteractiveState = {}, AuthoredState = {}, DialogState = {}, GlobalInteractiveState = {}>() => {
+  type InitMessage = IInitInteractive<InteractiveState, AuthoredState, DialogState, GlobalInteractiveState>;
+  const [ initMessage, setInitMessage ] = useState<InitMessage | null>(null);
+
+  useEffect(() => {
+    // useEffect callback can't be async.
+    (async () => {
+      setInitMessage(await client.getInitInteractiveMessage());
+    })();
+  }, []);
+
+  return initMessage;
+};
```

Comment:
One other hook I can see being useful is a `useSupportedFeatures()` hook which encapsulates the necessary `useEffect()` logic. I added a similar comment/question in the client PR.

Reply:
`useSupportedFeatures` actually is just a function that interactive calls and never waits for any results. So, I don't think we need a hook for it.

Actually, this brings another topic - do we really need it? So far LARA seems to accept multiple calls and in any order. But it might be a bit confusing for a developer - when this should be called, can it be called multiple times, etc etc. We mostly use `interactiveState`, `authoredState` and `aspectRatio`. Does LARA really need to know whether interactive supports interactive and authored state? Can't it just pass the sate if it's present? And always listen for incoming states. Aspect ratio could be moved to a separate function like `setHeight`. If there're some other features like global or linked state that require this upfront information for some reason, maybe we could split it into separate functions, so it's easier for hooks to encapsulate everything?

Reply:
The general intention was so the interactive could say what it supports and then the authoring and runtime system could adjust without needing to maintain external settings for the interactive. But in practice we haven't be able to implement it this way.  LARA now has settings to say if the interactive supports interactiveState, and whether it supports authoring.  The interactive state setting is used to determine whether it should block the page navigation to make sure the state of the interactive is saved. The authored state setting is used to determine if the authoring system should show the interactive in authored mode or it should just show some generic properties to customize the interactive.

But again these are settings in LARA instead of using the values from the supportedFeatures. This isn't great because it means when a developer changes an interactive we have to find all references to it and update their LARA settings. That is where the supportedFeatures came from. 

A new api we are adding is the ability for the interactive to support custom messages. This so plugins or interactives can send messages to the interactive. And that will be used by the teacher edition plugin to tell the multiple choice interactive to show its teacher-edition:correct overlay or teacher-edition:distractor overlay.  In order for the teacher edition authoring to function as it does now, it will need to know if the interactive supports these correct and distractor overlays. So the way we put this into the interactive types is through the supportedFeatures.  The interactive would say which of these custom messages it handles. The plugin authoring would have access to this and so then the teacher edition authoring could know if it is should show its correct and distractor tabs. 

However, all of this is really static meta-data about the interactive. It isn't something that will change while the interactive is running.  So if each interactive provided meta data next to its html page, then we could improve this approach. This is what we tried to do with the LARA plugins. 

However I don't think this approach with the plugins worked out so well. In particular it still requires an admin to go in and refresh the meta data. It has the benefit that the author can't set the meta data to be something totally wrong, and we don't need a UI for all of this metadata.  But it does mean the meta data can become stale because doesn't automatically get updated.

Anyhow, sorry I don't have a concrete answer, but hopefully having more people aware of the details will help us work out a solution.